README.md
[![Bazzite Project](https://img.shields.io/badge/Bazzite-project-blue.svg)](https://www.bazzite.com/docs/nativescript-version-tracking?utm_source=github&utm_medium=readme&utm_campaign=nativescript-version-tracking)
[![Travis](https://img.shields.io/travis/bazzite/nativescript-version-tracking/master.svg)](https://travis-ci.org/bazzite/nativescript-version-tracking)
[![version](https://img.shields.io/npm/v/nativescript-version-tracking.svg)](https://www.npmjs.com/package/nativescript-version-tracking)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://www.bazzite.com/docs/nativescript-version-tracking/license?utm_source=github&utm_medium=readme&utm_campaign=nativescript-version-tracking)

# NativeScript Version Tracking ![apple](https://cdn3.iconfinder.com/data/icons/picons-social/57/16-apple-32.png) ![android](https://cdn4.iconfinder.com/data/icons/logos-3/228/android-32.png) 

Track which versions of your NativeScript App, a user has previously installed.

## Installation

Run the following command from the root of your project:

```bash
$ tns plugin add nativescript-version-tracking
```

## Usage 

The best way to explore the usage of the plugin is to inspect the demo app in the plugin's root folder.

### Initialize the plugin

This plugin needs to be initialized when your app starts.

*TypeScript*

```typescript
import * as app from 'tns-core-modules/application';
import versionTracking from 'nativescript-version-tracking';

app.on('launch', () => {
    versionTracking.init();
});
```

*Javascript*

```javascript
var app = require('tns-core-modules/application');
var versionTracking = require('nativescript-version-tracking');

app.on('launch', function () {
    versionTracking.init();
});	
```

### Use the API

You can make use of the plugin whenever you want. For example, a user has launched several previous versions, and this is the first time he's launched the new version 2.0.1:

```javascript
versionTracking.isFirstLaunchEver;        // false
versionTracking.isFirstLaunchForVersion;  // true
versionTracking.isFirstLaunchForBuild;    // true

versionTracking.currentVersion;           // 2.0.1
versionTracking.previousVersion;          // 2.0.0
versionTracking.firstInstalledVersion;    // 1.0.0
versionTracking.versionHistory;           // [1.0.0, 1.0.1, 1.0.2, 2.0.0, 2.0.1]

versionTracking.currentBuild;             // 18
versionTracking.previousBuild;            // 15
versionTracking.firstInstalledBuild;      // 1
versionTracking.buildHistory;             // [1, 2, 3, 4, 5, 8, 9, 10, 11, 13, 15, 18]

versionTracking.firstLaunchForVersion('3.0.0')     // false
versionTracking.firstLaunchForBuild('20')          // false
```

## API


### Properties

| Property                | Default | Description                                                                                                                                                                             |
|-------------------------|:-------:|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| isFirstLaunchEver       |    -    | Check if this is the first time ever that the app is launched.                                                                                                                          |
| isFirstLaunchForVersion |    -    | Check if this is the first time the current version is being launched.                                                                                                                  |
| isFirstLaunchForBuild   |    -    | Check if this is the first time the current build is being launched.                                                                                                                    |
| currentVersion          |    -    | Returns the current version of the app.                                                                                                                                                 |
| previousVersion         |    -    | Returns the previous version of the app.                                                                                                                                                |
| firstInstalledVersion   |    -    | Returns the version which the user first installed the app.                                                                                                                             |
| versionHistory          |    []   | Returns a list of versions which the user has had installed, e.g. ['2.1', '3.5', '4.0', '4.1']. The List is ordered from the first version installed to (including) the current version |
| currentBuild            |    -    | Returns the current build of the app.                                                                                                                                                   |
| previousBuild           |    -    | Returns the previous build of the app.                                                                                                                                                  |
| firstInstalledBuild     |    -    | Returns the build which the user first installed the app.                                                                                                                               |
| buildHistory            |    []   | Returns a list of builds which the user has had installed, e.g. ['2100', '3500', '4000', '4100']. The List is ordered from the first build installed to (including) the current build.  |
### Methods

| Method                                         | Returns | Description                                                                                                                                                     |
|------------------------------------------------|:-------:|-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| init(versionsKey?: string, buildsKey?: string) |   void  | Initializes the plugin. Calling this method is required. A good place to call it is at the application onLaunch() method.                                       |
| firstLaunchForVersion(version: string)         | boolean | Check if this is the first launch for a particular version number. Useful if you want to execute some code for the first time launches of a particular version. |
| firstLaunchForBuild(build: string)             | boolean | Check if this is the first launch for a particular build number. Useful if you want to execute some code for the first time launches of a particular build.     |

## Support

- 🐞 For **Bug reports** or **Feature requests**, use the [Issues section][issues].
- 💬 For **questions**, go to [https://spectrum.chat/bazzite/open-source][spectrum-chat].
- 🚀 You may also want to [follow the company supporting this project on Twitter][twitter].

### Professional Support

This project is sponsored by [Bazzite][bazzite-website]. If you require assistance on your project(s), please contact us at [https://www.bazzite.com/contact][contact-page].

## Contributing

Please make sure to read the [Contributing Guide][contributing] before making a pull request.

## Code of Conduct

Everyone participating in this project is expected to agree to abide by the [Code of Conduct][code-of-conduct].

## License

Code released under the [MIT License][license-page].

---

Originally inspired by [VersionTrackingPlugin](https://github.com/colbylwilliams/VersionTrackingPlugin).


![](https://ga-beacon.appspot.com/UA-65885578-17/bazzite/nativescript-version-tracking?pixel)


[contributing]: https://www.bazzite.com/docs/nativescript-version-tracking/contributing?utm_source=github&utm_medium=readme&utm_campaign=nativescript-version-tracking
[code-of-conduct]: https://www.bazzite.com/open-source/code-of-conduct?utm_source=github&utm_medium=readme&utm_campaign=nativescript-version-tracking
[issues]: https://github.com/bazzite/nativescript-version-tracking/issues
[twitter]: https://twitter.com/BazziteTech
[spectrum-chat]: https://spectrum.chat/bazzite/login?r=https://spectrum.chat/bazzite/open-source
[bazzite-website]: https://www.bazzite.com?utm_source=github&utm_medium=readme&utm_campaign=nativescript-version-tracking
[contact-page]: https://www.bazzite.com/contact?utm_source=github&utm_medium=readme&utm_campaign=nativescript-version-tracking
[license-page]: https://www.bazzite.com/docs/nativescript-version-tracking/license?utm_source=github&utm_medium=readme&utm_campaign=nativescript-version-tracking


CHANGELOG.md
# Change Log

# [1.0.0](https://github.com/bazzite/vue-warehouse/compare/v0.1.0...v1.0.0) (2019-11-28)

## Refactor

* drop support for NativeScript v5 ([418aa6](https://github.com/bazzite/nativescript-version-tracking/commit/418aa67d210d7b386bad1d490367ff84c3dff5d0))
* use **default import** to import nativescript-version-tracking ([e95711](https://github.com/bazzite/nativescript-version-tracking/commit/e95711c4a771117ae71699e2808878ab21f11c66))

### BREAKING CHANGES

* minimum required NativeScript version is 6
* importing 'nativescript-version-tracking' on TypeScript should be done this way: `import versionTracking from 'nativescript-version-tracking';`


# [0.1.0](https://github.com/bazzite/nativescript-version-tracking/releases/tag/v0.1.0) (2018-12-03)

First Release

demo/app/tests/tests.js
var versionTracking = require('nativescript-version-tracking');

try {
    versionTracking.init()
} catch (error) {
    console.log('init error ->', error)
}

describe('histories', function() {
    it('versionHistory exists', function() {
        expect(versionTracking.versionHistory).toBeDefined();
    });

    it('versionHistory includes current version', function() {
        expect(versionTracking.versionHistory.includes(versionTracking.currentVersion)).toEqual(true);
    });

    it('buildHistory exists', function() {
        expect(versionTracking.buildHistory).toBeDefined();
    });

    it('buildHistory includes current build', function() {
        expect(versionTracking.buildHistory.includes(versionTracking.currentBuild)).toEqual(true);
    });
});

describe('version', function() {
    it('equals to 1.0', function() {
        expect(versionTracking.currentVersion).toEqual('1.0');
    });

    it('firstInstalledVersion equals to 1.0', function() {
        expect(versionTracking.firstInstalledVersion).toEqual('1.0');
    });

    it('previousVersion returns null', function() {
        expect(versionTracking.previousVersion).toEqual(null);
    });
});


describe('build', function() {
    it('equals to 10', function() {
        expect(versionTracking.currentBuild).toEqual('10');
    });

    it('firstInstalledBuild equals to 10', function() {
        expect(versionTracking.firstInstalledBuild).toEqual('10');
    });

    it('previousBuild returns null', function() {
        expect(versionTracking.previousBuild).toEqual(null);
    });
});


describe('methods', function() {
    it('firstLaunchForVersion returns false', function() {
        expect(versionTracking.firstLaunchForVersion('2349495586')).toEqual(false);
    });
    it('firstLaunchForBuild returns false', function() {
        expect(versionTracking.firstLaunchForBuild('2349495586')).toEqual(false);
    });
});

demo/app/main-view-model.ts
import { Observable } from 'tns-core-modules/data/observable';
import * as versionTracking from 'nativescript-version-tracking';

export class HelloWorldModel extends Observable {
  isFirstLaunchEver: boolean;
  isFirstLaunchForVersion: boolean;
  isFirstLaunchForBuild: boolean;

  currentVersion: string;
  previousVersion: string;
  firstInstalledVersion: string;
  versionHistory: string;

  currentBuild: string;
  previousBuild: string;
  firstInstalledBuild: string;
  buildHistory: string;

  firstLaunchForVersion: boolean;
  firstLaunchForBuild: boolean;

  constructor() {
    super();

    this.isFirstLaunchEver = versionTracking.isFirstLaunchEver;
    this.isFirstLaunchForVersion = versionTracking.isFirstLaunchForVersion;
    this.isFirstLaunchForBuild = versionTracking.isFirstLaunchForBuild;

    this.currentVersion = versionTracking.currentVersion;
    this.previousVersion = versionTracking.previousVersion;
    this.firstInstalledVersion = versionTracking.firstInstalledVersion;
    this.versionHistory = JSON.stringify(versionTracking.versionHistory);

    this.currentBuild = versionTracking.currentBuild;
    this.previousBuild = versionTracking.previousBuild;
    this.firstInstalledBuild = versionTracking.firstInstalledBuild;
    this.buildHistory = JSON.stringify(versionTracking.buildHistory);

    this.firstLaunchForVersion = versionTracking.firstLaunchForVersion('2.0.0');
    this.firstLaunchForBuild = versionTracking.firstLaunchForBuild('200');
  }
}


src/version-tracking.android.ts
import * as app from 'tns-core-modules/application';
import { versionTracking, init } from './version-tracking.common';

versionTracking.init = (versionsKey = 'tnsVersion', buildsKey = 'tnsBuild') => {
  if (versionTracking.initialized) {
    throw new Error('Version Tracking already initialized');
  }

  versionTracking.initialized = true;

  // current version and build
  const packageManager = android.content.pm.PackageManager;
  const packageInfo = app.android.context
    .getPackageManager()
    .getPackageInfo(app.android.context.getPackageName(), packageManager.GET_META_DATA);

  versionTracking.currentVersion = String(packageInfo.versionName);
  versionTracking.currentBuild = String(packageInfo.versionCode);

  // initialize
  init(versionTracking, versionsKey, buildsKey);
};

export default versionTracking;


demo/app/home/home-view-model.ts
import { Observable } from "tns-core-modules/data/observable";
import versionTracking from 'nativescript-version-tracking';

export class HomeViewModel extends Observable {
  isFirstLaunchEver: boolean;
  isFirstLaunchForVersion: boolean;
  isFirstLaunchForBuild: boolean;

  currentVersion: string;
  previousVersion: string;
  firstInstalledVersion: string;
  versionHistory: string;

  currentBuild: string;
  previousBuild: string;
  firstInstalledBuild: string;
  buildHistory: string;

  firstLaunchForVersion: boolean;
  firstLaunchForBuild: boolean;

  constructor() {
    super();

    this.isFirstLaunchEver = versionTracking.isFirstLaunchEver;
    this.isFirstLaunchForVersion = versionTracking.isFirstLaunchForVersion;
    this.isFirstLaunchForBuild = versionTracking.isFirstLaunchForBuild;

    this.currentVersion = versionTracking.currentVersion;
    this.previousVersion = versionTracking.previousVersion;
    this.firstInstalledVersion = versionTracking.firstInstalledVersion;
    this.versionHistory = JSON.stringify(versionTracking.versionHistory);

    this.currentBuild = versionTracking.currentBuild;
    this.previousBuild = versionTracking.previousBuild;
    this.firstInstalledBuild = versionTracking.firstInstalledBuild;
    this.buildHistory = JSON.stringify(versionTracking.buildHistory);

    this.firstLaunchForVersion = versionTracking.firstLaunchForVersion('2.0.0');
    this.firstLaunchForBuild = versionTracking.firstLaunchForBuild('200');
  }
}


src/version-tracking.common.ts
import * as appSettings from 'tns-core-modules/application-settings';

export interface IVersionTracking {
  initialized: boolean;

  /**
   * Check if this is the first time ever that the app is launched.
   */
  isFirstLaunchEver: boolean;

  /**
   * Check if this is the first time the current version is being launched.
   */
  isFirstLaunchForVersion: boolean;

  /**
   * Check if this is the first time the current build is being launched.
   */
  isFirstLaunchForBuild: boolean;


  /**
   * Returns the current version of the app.
   */
  currentVersion: string;

  /**
   * Returns the previous version of the app.
   */
  previousVersion: string;

  /**
   * Returns the version which the user first installed the app.
   */
  firstInstalledVersion: string;

  /**
   * Returns a list of versions which the user has had installed, e.g. ['2.1', '3.5', '4.0', '4.1']. The List is ordered from the first version installed to (including) the current version
   */
  versionHistory: Array<string>;


  /**
   * Returns the current build of the app.
   */
  currentBuild: string;

  /**
   * Returns the previous build of the app.
   */
  previousBuild: string;

  /**
   * Returns the build which the user first installed the app.
   */
  firstInstalledBuild: string;

  /**
   * Returns a list of builds which the user has had installed, e.g. ['2100', '3500', '4000', '4100']. The List is ordered from the first build installed to (including) the current build.
   */
  buildHistory: Array<string>;


  /**
   * Initializes the plugin. Calling this method is required. A good place to call it is at the application onLaunch() method.
   * @param versionsKey The versions key to be used in the Application Settings
   * @param buildsKey The builds version  to be used in the Application Settings
   */
  init(versionsKey?: string, buildsKey?: string): void;

  /**
   * Check if this is the first launch for a particular version number. Useful if you want to execute some code for the first time launches of a particular version.
   * @param version The particular version number
   * @return {boolean}
   */
  firstLaunchForVersion(version: string): boolean;

  /**
   * Check if this is the first launch for a particular build number. Useful if you want to execute some code for the first time launches of a particular build.
   * @param build This particular build number
   * @return {boolean}
   */
  firstLaunchForBuild(build: string): boolean;
}

export const init = (versionTracking: IVersionTracking, versionsKey: string, buildsKey: string) => {
  // load history
  let versionTrail: { [key: string]: Array<string> } = {};

  const oldVersionList: Array<string> = JSON.parse(appSettings.getString(versionsKey, null));
  const oldBuildList: Array<string> = JSON.parse(appSettings.getString(buildsKey, null));

  if (oldVersionList == null || oldBuildList == null) {
    versionTracking.isFirstLaunchEver = true;

    versionTrail[versionsKey] = [];
    versionTrail[buildsKey] = [];
  } else {
    versionTrail[versionsKey] = oldVersionList;
    versionTrail[buildsKey] = oldBuildList;

    versionTracking.isFirstLaunchEver = false;
  }

  // check if this version was previously launched
  if (versionTrail[versionsKey].indexOf(versionTracking.currentVersion) !== -1) {
    versionTracking.isFirstLaunchForVersion = false;
  } else {
    versionTracking.isFirstLaunchForVersion = true;

    versionTrail[versionsKey].push(versionTracking.currentVersion);
  }

  // check if this build was previously launched
  if (versionTrail[buildsKey].indexOf(versionTracking.currentBuild) !== -1) {
    versionTracking.isFirstLaunchForBuild = false;
  } else {
    versionTracking.isFirstLaunchForBuild = true;

    versionTrail[buildsKey].push(versionTracking.currentBuild);
  }

  // Previous Version
  const totalVersions = versionTrail[versionsKey].length;
  versionTracking.previousVersion =
    totalVersions >= 2 ? versionTrail[versionsKey][totalVersions - 2] : null;

  // First Installed Version
  versionTracking.firstInstalledVersion = versionTrail[versionsKey][0] || null;

  // Previous Build
  const totalBuilds = versionTrail[buildsKey].length;
  versionTracking.previousBuild =
    totalBuilds >= 2 ? versionTrail[buildsKey][totalBuilds - 2] : null;

  // first Installed Build
  versionTracking.firstInstalledBuild = versionTrail[buildsKey][0] || null;

  // histories
  versionTracking.versionHistory = versionTrail[versionsKey];
  versionTracking.buildHistory = versionTrail[buildsKey];

  // store the new version stuff
  appSettings.setString(versionsKey, JSON.stringify(versionTrail[versionsKey]));
  appSettings.setString(buildsKey, JSON.stringify(versionTrail[buildsKey]));
};

/**
 * Track which versions of your NativeScript App, a user has previously installed.
 * @module "nativescript-version-tracking"
 */ /** */
export const versionTracking: IVersionTracking = {
  initialized: false,

  isFirstLaunchEver: null,
  isFirstLaunchForVersion: null,
  isFirstLaunchForBuild: null,

  currentVersion: null,
  previousVersion: null,
  firstInstalledVersion: null,
  versionHistory: [],

  currentBuild: null,
  previousBuild: null,
  firstInstalledBuild: null,
  buildHistory: [],

  init: () => null,

  firstLaunchForVersion: (version: string): boolean => {
    if (versionTracking.currentVersion) {
      return versionTracking.currentVersion.toLowerCase() === version.toLowerCase() && versionTracking.isFirstLaunchForVersion;
    } else {
      return false;
    }
  },

  firstLaunchForBuild: (build: string): boolean => {
    if (versionTracking.currentBuild) {
      return versionTracking.currentBuild.toLowerCase() === build.toLowerCase() && versionTracking.isFirstLaunchForBuild;
    } else {
      return false;
    }
  }
};
